### DISCUSSION TITLE
How to set workspace-relative `outputs` with the relative package paths?

### DISCUSSION THREAD

**Discussion by niieani** [moonrepo/moon#1418] [Q&A] [answered]
Hi! and thank you again for this lovely piece of software, Miles! 🙇 

My directory structure (from workspace root) is as follows:

```
packages/my-package/package.json
packages/my-package/main.ts
```

The above gets built into =>

```
build/packages/my-package/package.json
build/packages/my-package/main.js
```

I then publish the packages from the `build` directory.
According to [moon documentation](https://moonrepo.dev/docs/concepts/file-pattern#project-relative), I can provide workspace-relative `outputs` by starting with `/`. But I don't see a way to reference the relative package path within the output using some kind of a dynamic placeholder. So the only way to actually cache the output of my build, is to provide the full path explicitly in the task definition. E.g. 

```
/build/packages/my-package/*
```

I'd like to be able to write the `outputs` of my task by providing some sort of a placeholder that contains the relative path to the project instead. For example:

```
/build/{projectDir}/*
```

And I would expect `{projectDir}` to automatically be replaced with `packages/my-package`.

Prior art: Lerna has a [configuration property called `directory`](https://lerna.js.org/docs/concepts/configuring-published-files#--directory) for this purpose, which for reference I have configured as: `{workspaceRoot}/build/{projectRoot}`

Is this possible with moon today? If not, is there a chance a feature like this might land in moon any time soon?

Also, if I had multiple build tasks that wrote to the same directory (but different files), would moon manage to cache them all this way? 
As an aside: It would be neat if we could provide the list of files written by a given process, from the output of that process, so moon knows exactly what to cache without having to provide `outputs`. Esp. for custom scripts this would be possible to achieve (since we have full control over what gets written to stdout/stderr).

🙇

**Comment by milesj** [comment, marked as answer]
@niieani Would the `$projectSource` token work here? https://moonrepo.dev/docs/concepts/token#projectsource This is the project path, without the workspace prefix.

> As an aside: It would be neat if we could provide the list of files written by a given process, from the output of that process, so moon knows exactly what to cache without having to provide outputs. Esp. for custom scripts this would be possible to achieve (since we have full control over what gets written to stdout/stderr).

This is a pretty neat idea. Kind of reminds of me GH actions a bit. I can look into this for the future.

**Reply by niieani** [reply]
`$projectSource` might just work! I will try it out! Thanks 🙇
